src/shogun/clustering/KMeansBase.cpp

```diff
@@ -354,17 +355,23 @@ SGMatrix<float64_t> CKMeansBase::kmeanspp()
 
 void CKMeansBase::init()
 {
-	max_iter=300;
-	k=8;
-	dimensions=0;
-	fixed_centers=false;
-	use_kmeanspp=false;
-	SG_ADD(&max_iter, "max_iter", "Maximum number of iterations", ParameterProperties::HYPER);
-	SG_ADD(&k, "k", "k, the number of clusters", ParameterProperties::HYPER);
+	max_iter = 300;
+	k = 8;
+	dimensions = 0;
+	fixed_centers = false;
+	use_kmeanspp = false;
+	SG_ADD(
+	    &max_iter, "max_iter", "Maximum number of iterations",
+	    ParameterProperties::HYPER);
+	SG_ADD_CONSTRAINT(
```

Comment:
One thing I don't like so much is that we need a new macro here.
The options were added as a part of the `SG_ADD` with this nice `SG_OPTION`.
Is it not possible to do a similar thing here?
`SG_ADD(&my_aram, "my_param", "descr", SG_CONSTRAINT(positive<>)`

Reply:
well I was thinking about this, and in an ideal world we would drop these macros. Instead there would be a `sg_add` method that has these overloads with variadic template arguments, and it would be type safe and easier to debug

Reply:
but yes, I think it should be possible to do that

Reply:
i agree with you actually. I think we wanted to keep it this way to keep the old framework working until we decide to kill it. But we should change to a method once the old param stuff goes (hopefully workshop? :D)

Reply:
I think constraints are optional so something like:
`add(...).constraint(positive())`
might be good.

Reply:
@lisitsyn if/when we move away from the SG_ADD macros?

Reply:
yeah seems so

Reply:
Maybe it is a good time to bury that macro then... we could also call the old param registrations in the new method?

Reply:
yup I don't see why not, I started working a bit on it, and it shouldn't be too hard even with all the new parameter properties.

Reply:
++++